### ISSUE TITLE
powerpc: _aligned_contig_cast_double_to_longdouble: Assertion `npy_is_aligned(dst, __builtin_offsetof (struct {char c; npy_longdouble v;}, v))' failed. Running unit tests for numpy

### ISSUE THREAD

**Issue by sandrotosi** [numpy/numpy#5550]
Hello,
with 1.9.2rc1 in Debian we got the following error running tests with python2.7 debug flavour when building the package on powerpc architecture:

```
test_dtype_from_dtype (test_multiarray.TestStats) ... python2.7-dbg: numpy/core/src/multiarray/lowlevel_strided_loops.c.src:820: _aligned_contig_cast_double_to_longdouble: Assertion `npy_is_aligned(dst, __builtin_offsetof (struct {char c; npy_longdouble v;}, v))' failed.
```

full build log at: https://buildd.debian.org/status/fetch.php?pkg=python-numpy&arch=powerpc&ver=1%3A1.9.2~rc1-1&stamp=1423426097

**Comment by juliantaylor** [comment]
I think I tested sparc in this rc, so powerpc is again different? ...
probably worth a short look whats wrong now, but if its too much effort supporting win32 over obscure arches should probably have preference.

**Comment by juliantaylor** [comment]
grr porterbox access ran out again ...

**Comment by sandrotosi** [comment]
can I help you somehow in getting back the access to the porterbox?

**Comment by juliantaylor** [comment]
no I have already requested it, it normally goes through in a few days.

**Comment by juliantaylor** [comment]
from what I can tell its due to malloc only providing 8 bytes alignment but to copy a complex long double we need 16 bytes. So the iterator buffer does not work.

A fix would be quite ugly, special code just for powerpc or just always do unaligned copy on that platform regardless of alignment.
Does anyone actually care about that platform? Else I'd opt for ignore the issue, users will have to turn on software fault handling for complex long doubles or fix the libc.

**Comment by juliantaylor** [comment]
glibc bug: https://sourceware.org/bugzilla/show_bug.cgi?id=6527

**Comment by charris** [comment]
On Mon, Feb 16, 2015 at 12:29 PM, Julian Taylor <notifications@github.com>
wrote:

> glibc bug: https://sourceware.org/bugzilla/show_bug.cgi?id=6527
>

They should just dump emacs ;)

Chuck

**Comment by sandrotosi** [comment]
In debian we came up with this patch, to remove the assert on ppc:

```
Description: disable asserts on ppc with broken malloc
 only longdouble affected will still work with softfaults
Author: Julian Taylor <jtaylor.debian@googlemail.com>
--- a/numpy/core/src/multiarray/lowlevel_strided_loops.c.src
+++ b/numpy/core/src/multiarray/lowlevel_strided_loops.c.src
@@ -7,6 +7,13 @@
  *
  * See LICENSE.txt for the license.
  */
+#if defined(__powerpc__) & !defined(__powerpc64__)
+#ifndef NDEBUG
+// disable asserts malloc broken
+// https://sourceware.org/bugzilla/show_bug.cgi?id=6527
+#define NDEBUG
+#endif
+#endif
 
 #define PY_SSIZE_T_CLEAN
 #include "Python.h"
```

maybe you can consider including it in numpy directly?

**Comment by charris** [comment]
@sandrotosi Is this still a problem?

**Comment by sandrotosi** [comment]
not this specific issue, but on powerpc several tests are failing: here is the build log for 1.14.3: https://buildd.debian.org/status/fetch.php?pkg=python-numpy&arch=powerpc&ver=1%3A1.14.3-2&stamp=1525580993&raw=0 (look for _running tests for_ to reach where the tests are running)

**Comment by mattip** [comment]
Can we close? The [debian builds](https://buildd.debian.org/status/package.php?p=numpy&suite=sid) look green but I don't see where they run tests.

**Comment by sandrotosi** [comment]
hehe the builds are green mostly because i cheated :) tests failures are not causing the package build to fail since otherwise we'd never be able to release numpy cause some of our most esoteric release archs (the one listed in white background at the top of the table at https://buildd.debian.org/status/package.php?p=numpy&suite=sid) almost always fails some tests

to get the build log for each arch, you can click the "Installed" label in the Status column; to jump to the tests results the best way i have is to search for the literal string `passed, ` which should appear only twice (since we are currently building for 3.10 and 3.11) and that'll have the pytest recap just above.

currently i see failures in:

https://buildd.debian.org/status/fetch.php?pkg=numpy&arch=armel&ver=1%3A1.24.1-2&stamp=1673226556&raw=0
https://buildd.debian.org/status/fetch.php?pkg=numpy&arch=armhf&ver=1%3A1.24.1-2&stamp=1673222642&raw=0
https://buildd.debian.org/status/fetch.php?pkg=numpy&arch=i386&ver=1%3A1.24.1-2&stamp=1673219434&raw=0
https://buildd.debian.org/status/fetch.php?pkg=numpy&arch=mips64el&ver=1%3A1.24.1-2&stamp=1673256220&raw=0
https://buildd.debian.org/status/fetch.php?pkg=numpy&arch=mipsel&ver=1%3A1.24.1-2&stamp=1673231309&raw=0

let me know if there's a better way to feed back tests failures during debian packages builds

**Comment by seberg** [comment]
There is the f2py test that also shows up on some M1 setups at least (gh-21765).  The mipsel one has a promotion error in `TestWhere.test_exotic` which seems strange (picks the wrong output dtype, that may be a real bug somewhere that just barely shows).

Mipsel64 has something that looks like SIMD failures, most look like not very critical ones but I am not sure. https://buildd.debian.org/status/fetch.
php?pkg=numpy&arch=mips64el&ver=1%3A1.24.1-2&stamp=1673256220&raw=0
```
_________________________ TestFPClass.test_fpclass[1] __________________________

self = <numpy.core.tests.test_umath.TestFPClass object at 0xfff01ed6c0>
stride = 1

    @pytest.mark.parametrize("stride", [-4,-2,-1,1,2,4])
    def test_fpclass(self, stride):
        arr_f64 = np.array([np.nan, -np.nan, np.inf, -np.inf, -1.0, 1.0, -0.0, 0.0, 2.2251e-308, -2.2251e-308], dtype='d')
        arr_f32 = np.array([np.nan, -np.nan, np.inf, -np.inf, -1.0, 1.0, -0.0, 0.0, 1.4013e-045, -1.4013e-045], dtype='f')
        nan     = np.array([True, True, False, False, False, False, False, False, False, False])
        inf     = np.array([False, False, True, True, False, False, False, False, False, False])
        sign    = np.array([False, True, False, True, True, False, True, False, False, True])
        finite  = np.array([False, False, False, False, True, True, True, True, True, True])
        assert_equal(np.isnan(arr_f32[::stride]), nan[::stride])
        assert_equal(np.isnan(arr_f64[::stride]), nan[::stride])
        assert_equal(np.isinf(arr_f32[::stride]), inf[::stride])
        assert_equal(np.isinf(arr_f64[::stride]), inf[::stride])
>       assert_equal(np.signbit(arr_f32[::stride]), sign[::stride])

arr_f32    = array([    nan,     nan,     inf,    -inf, -1.e+00,  1.e+00, -0.e+00,
        0.e+00,  1.e-45, -1.e-45], dtype=float32)
arr_f64    = array([         nan,          nan,          inf,         -inf,
       -1.0000e+000,  1.0000e+000, -0.0000e+000,  0.0000e+000,
        2.2251e-308, -2.2251e-308])
finite     = array([False, False, False, False,  True,  True,  True,  True,  True,
        True])
inf        = array([False, False,  True,  True, False, False, False, False, False,
       False])
nan        = array([ True,  True, False, False, False, False, False, False, False,
       False])
self       = <numpy.core.tests.test_umath.TestFPClass object at 0xfff01ed6c0>
sign       = array([False,  True, False,  True,  True, False,  True, False, False,
        True])
stride     = 1
```

This test seems to fail on 32bits when the requires memory check fails (or just succeeds) I presume, needs to check that as well.
```
    @requires_memory(6 * 1024**3)
    def test_identityless_reduction_huge_array(self):
        # Regression test for gh-20921 (copying identity incorrectly failed)
>       arr = np.zeros((2, 2**31), 'uint8')
E       ValueError: Maximum allowed dimension exceeded
```
That `TestFPClass.test_fpclass` error shows up there as well.  It may need some looking into or marking with xfail...

The f2py kind test fails all over @HaoZeke is this important, or can we somehow fix this up generically?

The spurious exception test also fails on armvhf, maybe a similar issue as on M1 which I think was fixed in 1.24.2 (pretty harmless one also).

Overall, we are at least a lot better than in 2015, I think :).  I will try to have a quick look at the `where` failure at least, but seems like it may be hard to track down.

**Comment by sandrotosi** [comment]
closing outdated issues i reported long ago
